In [ ]:
import numpy as np

Q1) a) Reverse the NumPy array: arr = np.array([1, 2, 3, 6, 4, 5])

In [ ]:
arr=np.array([1, 2, 3, 6, 4, 5])
reversed_array=np.flip(arr)
print(reversed_array)

[5 4 6 3 2 1]


OUTPUT:[5 4 6 3 2 1]


(b) Flatten the NumPy arr: array1 = np.array([[1, 2, 3], [2, 4, 5], [1, 2, 3]]) using any two NumPy in-built methods


In [ ]:
arr=np.array([[1, 2, 3], [2, 4, 5], [1, 2, 3]])
flatten_array=arr.flatten()
print(flatten_array)

[1 2 3 2 4 5 1 2 3]


OUTPUT:[1 2 3 2 4 5 1 2 3]


(c) Compare the following numpy arrays:

arr1 = np.array([[1, 2], [3, 4]])

arr2 = np.array([[1, 2], [3, 4]])

In [ ]:
arr1 = np.array([[1, 2], [3, 4]])

arr2 = np.array([[1, 2], [3, 4]])
comparison = np.array_equal(arr1, arr2)
print(comparison)

True


OUTPUT:True

d) Find the most frequent value and their indice(s) in the following arrays:    
i.)x = np.array([1,2,3,4,5,1,2,1,1,1]).  
ii.)y = np.array([1, 1, 1, 2, 3, 4, 2, 4, 3, 3, ])


In [ ]:
x = np.array([1, 2, 3, 4, 5, 1, 2, 1, 1, 1])

most_frequent = np.bincount(x).argmax()

indices = np.where(x == most_frequent)[0]

print("Most frequent value:", most_frequent)
print("Indices where it appears:", indices)

from scipy import stats

y = np.array([1, 1, 1, 2, 3, 4, 2, 4, 3, 3])

values, counts = np.unique(y, return_counts=True)
max_count = counts.max()
most_frequent_values = values[counts == max_count]


for val in most_frequent_values:
    indices = np.where(y == val)[0]
    print(f"Most frequent value: {val}, Indices: {indices}")


Most frequent value: 1
Indices where it appears: [0 5 7 8 9]
Most frequent value: 1, Indices: [0 1 2]
Most frequent value: 3, Indices: [4 8 9]


OUTPUT:
Most frequent value: 1


Indices where it appears: [0 5 7 8 9]


Most frequent value: 1, Indices: [0 1 2]


Most frequent value: 3, Indices: [4 8 9]

e)For the array array gfg = np.matrix('[4, 1, 9; 12, 3, 1; 4, 5, 6]'), find.  
i) Sum of all elements.  
ii)Sum of all elements row-wise.  
iii)Sum of all elements column-wise

In [ ]:
gfg = np.matrix('[4, 1, 9; 12, 3, 1; 4, 5, 6]')
sum_all = np.sum(gfg)
sum_rowwise = np.sum(gfg, axis=1)
sum_colwise = np.sum(gfg, axis=0)
print('sum of all elemnets:',sum_all)
print('Sum of all elements row-wise:',sum_rowwise)
print('Sum of all elements column-wise:',sum_colwise)

sum of all elemnets: 45
Sum of all elements row-wise: [[14]
 [16]
 [15]]
Sum of all elements column-wise: [[20  9 16]]


OUTPUT:sum of all elemnets: 45.  
Sum of all elements row-wise: [[14]
 [16]
 [15]]    
Sum of all elements column-wise: [[20  9 16]]

(f) For the matrix: n_array = np.array([[55, 25, 15],[30, 44, 2],[11, 45, 77]]), find.  
i)Sum of diagonal elements   
ii)Eigen values of matrix   
iii)Eigen vectors of matrix   
iv)Inverse of matrix   
 v)Determinant of matrix

In [ ]:
n_array = np.array([[55, 25, 15],[30, 44, 2],[11, 45, 77]])
diag_sum = np.trace(n_array)
eig_vals, eig_vecs = np.linalg.eig(n_array)
inv_matrix = np.linalg.inv(n_array)
det_matrix = np.linalg.det(n_array)
print('Sum of diagonal elements:',diag_sum)
print('Eigen values of matrix:',eig_vals)
print('Eigen Vectors:',eig_vecs)
print('Inversion of matrix:',inv_matrix)
print('Determinant of matrix:',det_matrix)

Sum of diagonal elements: 176
Eigen values of matrix: [98.16835147 28.097044   49.73460452]
Eigen Vectors: [[ 0.4574917   0.34637121 -0.15017693]
 [ 0.28447814 -0.72784061 -0.4852124 ]
 [ 0.84248058  0.59184038  0.8614034 ]]
Inversion of matrix: [[ 0.02404141 -0.00911212 -0.00444671]
 [-0.01667882  0.02966905  0.0024785 ]
 [ 0.00631287 -0.01603732  0.01217379]]
Determinant of matrix: 137180.0000000001


OUTPUT:Sum of diagonal elements: 176
Eigen values of matrix: [98.16835147 28.097044   49.73460452]    
Eigen Vectors: [[ 0.4574917   0.34637121 -0.15017693]
 [ 0.28447814 -0.72784061 -0.4852124 ]
 [ 0.84248058  0.59184038  0.8614034 ]].    
Inversion of matrix: [[ 0.02404141 -0.00911212 -0.00444671]
 [-0.01667882  0.02966905  0.0024785 ]
 [ 0.00631287 -0.01603732  0.01217379]].    
Determinant of matrix: 137180.0000000001

G)Multiply the following matrices and also find covariance between matrices using NumPy:   
i. p = [[1, 2], [2, 3]].
q = [[4, 5], [6, 7]].  
ii. p = [[1, 2], [2, 3], [4, 5]]
q = [[4, 5, 1], [6, 7, 2]]

In [ ]:
# (i.)
p1 = np.array([[1, 2], [2, 3]])
q1 = np.array([[4, 5], [6, 7]])

product1 = np.dot(p1, q1)
print("Matrix Multiplication (p1 x q1):\n", product1)

combined1 = np.vstack((p1.flatten(), q1.flatten()))
cov_matrix1 = np.cov(combined1)
print("\nCovariance matrix between p1 and q1:\n", cov_matrix1)


# (ii.)
p2 = np.array([[1, 2], [2, 3], [4, 5]])
q2 = np.array([[4, 5, 1], [6, 7, 2]])


product2 = np.dot(p2, q2)
print("\nMatrix Multiplication (p2 x q2):\n", product2)


combined2 = np.vstack((p2.flatten(), q2.flatten()))
cov_matrix2 = np.cov(combined2)
print("\nCovariance matrix between p2 and q2:\n", cov_matrix2)

Matrix Multiplication (p1 x q1):
 [[16 19]
 [26 31]]

Covariance matrix between p1 and q1:
 [[0.66666667 1.        ]
 [1.         1.66666667]]

Matrix Multiplication (p2 x q2):
 [[16 19  5]
 [26 31  8]
 [46 55 14]]

Covariance matrix between p2 and q2:
 [[2.16666667 0.23333333]
 [0.23333333 5.36666667]]


OUTPUT:Matrix Multiplication (p1 x q1):
 [[16 19]
 [26 31]]

Covariance matrix between p1 and q1:
 [[0.66666667 1.        ]
 [1.         1.66666667]]

Matrix Multiplication (p2 x q2):
 [[16 19  5]
 [26 31  8]
 [46 55 14]]

Covariance matrix between p2 and q2:
 [[2.16666667 0.23333333]
 [0.23333333 5.36666667]]

(h) For the matrices:   

x = np.array([[2, 3, 4], [3, 2, 9]]);

y = np.array([[1, 5, 0], [5, 10, 3]]),

find inner, outer and cartesian product?



In [ ]:


x = np.array([[2, 3, 4], [3, 2, 9]])
y = np.array([[1, 5, 0], [5, 10, 3]])

inner_product = np.sum(x * y)


outer_product = np.outer(x, y)


cartesian_product = np.array(np.meshgrid(x, y)).T.reshape(-1, x.shape[1] + y.shape[1])

print("Inner Product:", inner_product)
print("\nOuter Product:\n", outer_product)
print("\nCartesian Product:\n", cartesian_product)

Inner Product: 79

Outer Product:
 [[ 2 10  0 10 20  6]
 [ 3 15  0 15 30  9]
 [ 4 20  0 20 40 12]
 [ 3 15  0 15 30  9]
 [ 2 10  0 10 20  6]
 [ 9 45  0 45 90 27]]

Cartesian Product:
 [[ 2  1  2  5  2  0]
 [ 2  5  2 10  2  3]
 [ 3  1  3  5  3  0]
 [ 3  5  3 10  3  3]
 [ 4  1  4  5  4  0]
 [ 4  5  4 10  4  3]
 [ 3  1  3  5  3  0]
 [ 3  5  3 10  3  3]
 [ 2  1  2  5  2  0]
 [ 2  5  2 10  2  3]
 [ 9  1  9  5  9  0]
 [ 9  5  9 10  9  3]]


OUTPUT

Inner Product: 79

Outer Product:
 [[ 2 10  0 10 20  6]
 [ 3 15  0 15 30  9]
 [ 4 20  0 20 40 12]
 [ 3 15  0 15 30  9]
 [ 2 10  0 10 20  6]
 [ 9 45  0 45 90 27]]

Cartesian Product:
 [[ 2  1  2  5  2  0]
 [ 2  5  2 10  2  3]
 [ 3  1  3  5  3  0]
 [ 3  5  3 10  3  3]
 [ 4  1  4  5  4  0]
 [ 4  5  4 10  4  3]
 [ 3  1  3  5  3  0]
 [ 3  5  3 10  3  3]
 [ 2  1  2  5  2  0]
 [ 2  5  2 10  2  3]
 [ 9  1  9  5  9  0]
 [ 9  5  9 10  9  3]]

Q2

a) For the array: array = np.array([[1, -2, 3],[-4, 5, -6]]).   
i. Find element-wise absolute value.    
ii. Find the 25th, 50th, and 75th percentile of flattened array, for each column, for each row.    
iii. Mean, Median and Standard Deviation of flattened array, of each column, and each row

In [ ]:

array = np.array([[1, -2, 3], [-4, 5, -6]])



abs_array = np.abs(array)
print("Element-wise Absolute Value:\n", abs_array)



percentiles_flat = np.percentile(array, [25, 50, 75])
print("\nPercentiles of flattened array (25th, 50th, 75th):", percentiles_flat)


percentiles_cols = np.percentile(array, [25, 50, 75], axis=0)
print("Percentiles for each column:\n", percentiles_cols)


percentiles_rows = np.percentile(array, [25, 50, 75], axis=1)
print("Percentiles for each row:\n", percentiles_rows)



mean_flat = np.mean(array)
median_flat = np.median(array)
std_flat = np.std(array)
print("\nFlattened Array:")
print("Mean:", mean_flat)
print("Median:", median_flat)
print("Standard Deviation:", std_flat)


mean_col = np.mean(array, axis=0)
median_col = np.median(array, axis=0)
std_col = np.std(array, axis=0)
print("\nColumn-wise:")
print("Mean:", mean_col)
print("Median:", median_col)
print("Standard Deviation:", std_col)


mean_row = np.mean(array, axis=1)
median_row = np.median(array, axis=1)
std_row = np.std(array, axis=1)
print("\nRow-wise:")
print("Mean:", mean_row)
print("Median:", median_row)
print("Standard Deviation:", std_row)

Element-wise Absolute Value:
 [[1 2 3]
 [4 5 6]]

Percentiles of flattened array (25th, 50th, 75th): [-3.5 -0.5  2.5]
Percentiles for each column:
 [[-2.75 -0.25 -3.75]
 [-1.5   1.5  -1.5 ]
 [-0.25  3.25  0.75]]
Percentiles for each row:
 [[-0.5 -5. ]
 [ 1.  -4. ]
 [ 2.   0.5]]

Flattened Array:
Mean: -0.5
Median: -0.5
Standard Deviation: 3.8622100754188224

Column-wise:
Mean: [-1.5  1.5 -1.5]
Median: [-1.5  1.5 -1.5]
Standard Deviation: [2.5 3.5 4.5]

Row-wise:
Mean: [ 0.66666667 -1.66666667]
Median: [ 1. -4.]
Standard Deviation: [2.05480467 4.78423336]


OUTPUT:Element-wise Absolute Value:
 [[1 2 3]
 [4 5 6]]

Percentiles of flattened array (25th, 50th, 75th): [-3.5 -0.5  2.5]
Percentiles for each column:
 [[-2.75 -0.25 -3.75]
 [-1.5   1.5  -1.5 ]
 [-0.25  3.25  0.75]]
Percentiles for each row:
 [[-0.5 -5. ]
 [ 1.  -4. ]
 [ 2.   0.5]]

Flattened Array:
Mean: -0.5
Median: -0.5
Standard Deviation: 3.8622100754188224

Column-wise:
Mean: [-1.5  1.5 -1.5]
Median: [-1.5  1.5 -1.5]
Standard Deviation: [2.5 3.5 4.5]

Row-wise:
Mean: [ 0.66666667 -1.66666667]
Median: [ 1. -4.]
Standard Deviation: [2.05480467 4.78423336]

(b) For the array: a = np.array([-1.8, -1.6, -0.5, 0.5,1.6, 1.8, 3.0]).

Find floor, ceiling
and truncated value, rounded value

In [ ]:

a = np.array([-1.8, -1.6, -0.5, 0.5, 1.6, 1.8, 3.0])

floor_vals = np.floor(a)
print("Floor values: ", floor_vals)


ceil_vals = np.ceil(a)
print("Ceiling values: ", ceil_vals)

trunc_vals = np.trunc(a)
print("Truncated values: ", trunc_vals)


rounded_vals = np.round(a)
print("Rounded values: ", rounded_vals)

Floor values:  [-2. -2. -1.  0.  1.  1.  3.]
Ceiling values:  [-1. -1. -0.  1.  2.  2.  3.]
Truncated values:  [-1. -1. -0.  0.  1.  1.  3.]
Rounded values:  [-2. -2. -0.  0.  2.  2.  3.]


OUTPUT:

Floor values:  [-2. -2. -1.  0.  1.  1.  3.]


Ceiling values:  [-1. -1. -0.  1.  2.  2.  3.]


Truncated values:  [-1. -1. -0.  0.  1.  1.  3.]


Rounded values:  [-2. -2. -0.  0.  2.  2.  3.]


Q3)

(a) For the array: array = np.array([10, 52, 62, 16, 16, 54, 453]),

find:- i. Sorted array


ii. Indices of sorted array

iii. 4 smallest elements

iv. 5 largest elements

In [ ]:
array=np.array([10,52, 62, 16, 16, 54, 453])
sorted_array = np.sort(array)
print("Sorted array:", sorted_array)


sorted_indices = np.argsort(array)
print("Indices of sorted array:", sorted_indices)


smallest_4 = np.sort(array)[:4]
print("4 smallest elements:", smallest_4)


largest_5 = np.sort(array)[-5:][::-1]
print("5 largest elements:", largest_5)

Sorted array: [ 10  16  16  52  54  62 453]
Indices of sorted array: [0 3 4 1 5 2 6]
4 smallest elements: [10 16 16 52]
5 largest elements: [453  62  54  52  16]


OUTPUT:Sorted array: [ 10  16  16  52  54  62 453]


Indices of sorted array: [0 3 4 1 5 2 6]


4 smallest elements: [10 16 16 52]


5 largest elements: [453  62  54  52  16]

b) For the array: array = np.array([1.0, 1.2, 2.2, 2.0, 3.0, 2.0]),

find i. Integer elements only


ii. Float elements only

In [ ]:
array = np.array([1.0, 1.2, 2.2, 2.0, 3.0, 2.0])


integer_elements = array[array == np.floor(array)]
print("Integer elements:", integer_elements)

float_elements = array[array != np.floor(array)]
print("Float elements:", float_elements)

Integer elements: [1. 2. 3. 2.]
Float elements: [1.2 2.2]


OUTPUT:Integer elements: [1. 2. 3. 2.]
Float elements: [1.2 2.2]

Q4)



(a) Write a function named img_to_array(path) that reads an image from a specified path and save it as text file on local machine? (Note: use separate cases for RGB and Grey Scale images)


In [28]:
import cv2
import numpy as np

def img_to_array(path):
    image = cv2.imread(path, cv2.IMREAD_UNCHANGED)
    if len(image.shape) == 2:
        np.savetxt("image_arr.txt", image, fmt='%d')
        print(" Grayscale image detected and saved as 'image_arr.txt'")

    elif len(image.shape) == 3 and image.shape[2] == 3:
        reshaped = image.reshape(-1, 3)
        np.savetxt("image_arr.txt", reshaped, fmt='%d')
        print("Color image saved as 'image_arr.txt'")

    else:
        print(" Unsupported image format.")

img_to_array("image.jpg")

Color image saved as 'image_arr.txt'
